src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -492,6 +492,355 @@ TEST(Aktualizr, FullWithUpdatesNeedReboot) {
+    // moreover it takes version out of the stored metadata but not from successfully installed target
+    // see sotauptaneclient.cc::413
+    ASSERT_EQ(update_res.status, result::UpdateStatus::kError);
+    ASSERT_FALSE(aktualizr.uptane_client_->hasPendingUpdates());
+    ASSERT_TRUE(event_hdlr.checkRecievedEvents({"UpdateCheckComplete"}));
+    ASSERT_TRUE(http_server_mock->checkRecievedReports({"EcuInstallationCompleted"}));
+    ASSERT_FALSE(http_server_mock->wasInstallSuccessfull());
+
+    data::InstallationResult dev_installation_res;
+    std::string report;
+    std::string correlation_id;
+
+    // `device_installation_result` and `ecu_installation_results` are cleared
+    // at finalizeAfterReboot()->putManifestSimple() once a device manifest is successfully sent to a server
+    ASSERT_FALSE(storage->loadDeviceInstallationResult(&dev_installation_res, &report, &correlation_id));
+
+    // it's used to return `true` even if there is no any record in DB
+    // of the uptane cycle just after sending manifest
+    // made it consistent with loadDeviceInstallationResult
+    std::vector<std::pair<Uptane::EcuSerial, data::InstallationResult>> ecu_installation_res;
+    EXPECT_FALSE(storage->loadEcuInstallationResults(&ecu_installation_res));
+
+    // verify currently installed version
+    std::vector<Uptane::Target> installed_versions;
+    size_t current_version{0};  // why it's not boolean
+    size_t pending_version{0};  // why it's not boolean
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+    ASSERT_EQ(installed_versions.size(), 1);
+    // why valid , BTW this `valid` value is neither read from DB nor calculated and it's true by default ???
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+
+    current_version = 0;  // why it's not boolean
+    pending_version = 0;  // why it's not boolean
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(secondary_ecu_id, &installed_versions, &current_version, &pending_version));
+    ASSERT_EQ(installed_versions.size(), 1);
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+  }
+}
+
+/*
+ * Initialize -> UptaneCycle -> download updates -> fail installation
+ *
+ * Verifies whether the uptane client is not at pending state after installation failure
+ *
+ * Checks actions:
+ *
+ * - [x] Update is not in pending state anymore after failed installation
+ * - [x] Store negative device installation result when an ECU installation failed
+ */
+TEST(Aktualizr, InstallationFailure) {
+  TemporaryDirectory temp_dir;
+  auto http_server_mock = std::make_shared<HttpInstallationFailed>(temp_dir.Path());
+  Config conf = UptaneTestCommon::makeTestConfig(temp_dir, http_server_mock->tls_server);
+  conf.uptane.polling_sec = 0;
```

Comment:
Why set `polling_sec`? Shouldn't be necessary if you aren't using `UptaneCycle`.

Reply:
yes, you're right.